File: python/problem14.py
def genCollatzNumber(n):
    if n%2 == 0:
        return n/2
    else:
        return 3*n + 1

def countCollatzSequence(num):
    count=1
    while num != 1:
        num = genCollatzNumber(num)
        count += 1

    return count

def findLongestCollatzSequence(limit):
    maxCount = 1
    maxNum = 1
    for i in range(2,limit):
        result = countCollatzSequence(i)
        if result > maxCount:
            maxCount = result
            maxNum = i

    return maxNum

if __name__=='__main__':
    print(findLongestCollatzSequence(14))

File: go/problem14.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(FindLongestCollatzSequence(14))
}

func genCollatzSequence(n int) int {
	if n%2 == 0 {
		return n / 2
	} else {
		return 3*n + 1
	}
}

func FindLongestCollatzSequence(limit int) int {

	var maxCount, maxNumber int

	for number := 2; number <= limit; number++ {

		result := countCollatzSequenceToOne(number)
		if result > maxCount {
			maxCount = result
			maxNumber = number
		}

	}

	return maxNumber
}

func countCollatzSequenceToOne(num int) int {

	count := 1
	// fmt.Printf("genCollatzSequence is ")
	for true {

		num = genCollatzSequence(num)
		count++
		// fmt.Printf("%d ", num)
		if num == 1 {
			break
		}

	}
	// fmt.Printf("\nCount is %d \n", count)
	return count

}
